### PR TITLE
Add rust native implementation of product

### PR DESCRIPTION
**Pull request by adeebahmed**
This PR contains rust native implementations of the data models and functions associated, in product

### CODE REVIEW

File: sdk/src/protocol/product/payload.rs [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+use crate::protocol::product::state::PropertyDefinition;
+use crate::protos;
+use crate::protos::{
+    FromBytes, FromNative, FromProto, IntoBytes, IntoNative, IntoProto, ProtoConversionError,
+};
+
+/// Native implementation for ProductPayload_Action
+#[derive(Debug, Clone, PartialEq)]
+pub enum Action {
+    ProductCreate,
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Instead of following the proto exactly like, we can take advantage of rust enums so this Action enum can be 
```
pub enum Action {
   ProductCreate(ProductCreateAction),
   ProductUpdate(ProductUpdateAction),
   ProductDelete(ProducDeleteAction)
}
```
And then the Payload would be something like
```
pub struct ProductPayload  
    action: Action,
}
```

File: sdk/src/protocol/product/payload.rs [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+    }
+}
+
+impl IntoProto<protos::product_payload::ProductPayload_Action> for Action {}
+impl IntoNative<Action> for protos::product_payload::ProductPayload_Action {}
+
+// Rust native implementation for ProductPayload
+#[derive(Debug, Clone, PartialEq)]
+pub struct ProductPayload {
+    action: Action,
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Missing timestamp which you included in sdk/protos/

File: sdk/src/protocol/product/payload.rs [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+
+        let identifier = self.identifier.ok_or_else(|| {
+            ProductCreateBuildError::MissingField("'identifier' field is required".to_string())
+        })?;
+
+        let owner = self.owner.ok_or_else(|| {
+            ProductCreateBuildError::MissingField("'owner' field is required".to_string())
+        })?;
+
+        // Properties fields can be empty
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: You still need to provide the properties to the ProductCreateAction.

File: sdk/src/protocol/product/payload.rs [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+
+impl std::fmt::Display for ProductUpdateBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ProductUpdateBuildError::MissingField(ref s) => write!(f, "MissingField: {}", s),
+        }
+    }
+}
+
+/// Builder used to create a ProductPayload
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: ProductPayload -> ProductUpdateBuilder

File: sdk/src/protocol/product/payload.rs [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+            } else {
+                return Err(ProductUpdateBuildError::MissingField(
+                    "'properties' field is required".to_string(),
+                ));
+            }
+        };
+
+        Ok(ProductUpdateAction {
+            product_type,
+            identifier,
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Where is the identifier defined?

File: sdk/src/protocol/product/payload.rs [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+
+impl std::fmt::Display for ProductDeleteBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ProductDeleteBuildError::MissingField(ref s) => write!(f, "MissingField: {}", s),
+        }
+    }
+}
+
+/// Builder used to create a ProductPayload
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: ProductDeleteBuilder

File: sdk/src/protocol/product/payload.rs [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+        let identifier = self.identifier.ok_or_else(|| {
+            ProductDeleteBuildError::MissingField("'identifier is required".to_string())
+        })?;
+
+        Ok(ProductDeleteAction {
+            product_type,
+            identifier,
+        })
+    }
+}
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Missing tests

File: sdk/src/protocol/product/payload.rs (hunk 1 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

File: sdk/src/protocol/product/payload.rs (hunk 2 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

File: sdk/src/protocol/product/payload.rs (hunk 3 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

File: sdk/src/protocol/product/payload.rs (hunk 4 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

File: sdk/src/protocol/product/payload.rs (hunk 5 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

**Comment by peterschwarz** [comment]
Why didn't that compilation failure get caught by Jenkins?

**Comment by agunde406** [comment]
Since product is not added the protocol mod.rs file the product protocols are not being included.

File: sdk/src/protocol/product/payload.rs (hunk 6 above)
Comment (adeebahmed) [reply sdk/src/protocol/product/payload.rs]: ✅

**Comment by pa3ng** [comment]
Looks like adding in the rust native implementations for grid product is causing some sections of code in the smart contract to break; will update the product smart contract within this PR.

**Comment by pa3ng** [comment]
This open issue is causing the Jenkins build to break: https://github.com/rust-lang/rust/issues/62562

**tl;dr:* serde does not build with the latest version of nightly

**Comment by eloaverona** [comment]
@pa3ng It seems the issue was fixed in the newest nightly. I restarted your build in Jenkins and it passed :)

**Comment by pa3ng** [comment]
> This does not compile. You need to add `pub mod product;` to src/protocol/mod.rs and fix the compilation errors.

@agunde406 ✅

File: contracts/product/src/handler.rs [hunk 8]
```diff
@@ -100,47 +107,37 @@ impl ProductTransactionHandler {
             )));
         }
 
-        // Check that the agent has the pike permission "can_create_product" for the organization
-        check_permission(perm_checker, signer, "can_create_product")?;
```
Comment (adeebahmed) [inline contracts/product/src/handler.rs]: This check should not be removed

File: contracts/product/src/handler.rs [hunk 9]
```diff
@@ -222,31 +226,43 @@ impl ProductTransactionHandler {
...
         // Check if the agent updating the product is part of the organization associated with the product
-        if product.get_owner() != agent.org_id() {
+        if product.owner() != agent.org_id() {
             return Err(ApplyError::InvalidTransaction(
                 "Invalid organization for the agent submitting this transaction".to_string(),
             ));
         }
 
-        // Check that the agent has the pike permission "can_update_product" for the organization
-        check_permission(perm_checker, signer, "can_update_product")?;
```
Comment (adeebahmed) [inline contracts/product/src/handler.rs]: keep this check as well

File: contracts/product/src/handler.rs [hunk 10]
```diff
@@ -281,16 +290,18 @@ impl ProductTransactionHandler {
...
         // Check that the owner of the products organization is the same as the agent trying to delete the product
-        if product.get_owner() != agent.org_id() {
+        if product.owner() != agent.org_id() {
             return Err(ApplyError::InvalidTransaction(
                 "Invalid organization for the agent submitting this transaction".to_string(),
             ));
         }
 
-        // Check that the agent deleting the product has the "can_delete_product" permission for the organization
-        check_permission(perm_checker, signer, "can_delete_product")?;
```
Comment (adeebahmed) [inline contracts/product/src/handler.rs]: same here

File: contracts/product/src/handler.rs (hunk 8 above)
Comment (pa3ng) [reply contracts/product/src/handler.rs]: ✅

File: contracts/product/src/handler.rs (hunk 9 above)
Comment (pa3ng) [reply contracts/product/src/handler.rs]: ✅

File: contracts/product/src/handler.rs (hunk 10 above)
Comment (pa3ng) [reply contracts/product/src/handler.rs]: ✅

**Comment by pa3ng** [comment]
After talking with @peterschwarz about not creating a central, shared test package, I plan on adding the smart contract unit tests in this PR. Let me know if this is desired or not.

**TODO**
- [x] Add rust native implementations to product
- [x] Update product smart contract to use rust native implementations of product
- [ ] Add unit tests

File: sdk/protos/product_payload.proto [hunk 11]
```diff
@@ -37,36 +37,24 @@ message ProductPayload {
 }
 
 message ProductCreateAction {
-    enum ProductType {
```
Comment (agunde406) [inline sdk/protos/product_payload.proto]: Please add a comment to the commit message explaining this change

File: sdk/src/protocol/product/payload.rs [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+            ),
+            ProductPayload_Action::PRODUCT_UPDATE => Action::ProductUpdate(
+                ProductUpdateAction::from_proto(payload.get_product_update().clone())?,
+            ),
+            ProductPayload_Action::PRODUCT_DELETE => Action::ProductDelete(
+                ProductDeleteAction::from_proto(payload.get_product_delete().clone())?,
+            ),
+            ProductPayload_Action::UNSET_ACTION => {
+                return Err(ProtoConversionError::InvalidTypeError(
+                    "Cannot convert ProductPayload_Action with type unset.".to_string(),
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Remove period at the end of the error

File: sdk/src/protocol/product/payload.rs [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+
+impl std::fmt::Display for ProductCreateBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ProductCreateBuildError::MissingField(ref s) => write!(f, "MissingField: {}", s),
+        }
+    }
+}
+
+/// Builder used to create a ProductPayload
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: ProductPaylaod -> ProductCreateBuilder

File: sdk/src/protocol/product/state.rs [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+
+impl StdError for ProductBuildError {
+    fn description(&self) -> &str {
+        match *self {
+            ProductBuildError::MissingField(ref msg) => msg,
+            ProductBuildError::EmptyVec(ref msg) => msg,
+        }
+    }
+
+    fn cause(&self) -> Option<&StdError> {
```
Comment (peterschwarz) [inline sdk/src/protocol/product/state.rs]: `cause` is deprecated - you should use `source` for these:  https://doc.rust-lang.org/std/error/trait.Error.html#method.source

File: sdk/src/protocol/product/payload.rs [hunk 15]
```diff
@@ -0,0 +1,711 @@
...
+        self
+    }
+    pub fn with_properties(mut self, value: Vec<PropertyValue>) -> Self {
+        self.properties = Some(value);
+        self
+    }
+    pub fn build(self) -> Result<ProductCreateAction, BuilderError> {
+        let product_type = self
+            .product_type
+            .ok_or_else(|| BuilderError::MissingField("product_type".into()))?;
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: can you make these match the other errors " 'product_type' field is required" etc

File: sdk/src/protocol/product/payload.rs [hunk 16]
```diff
@@ -0,0 +1,711 @@
...
+    }
+
+    pub fn with_properties(mut self, properties: Vec<PropertyValue>) -> ProductUpdateBuilder {
+        self.properties = properties;
+        self
+    }
+
+    pub fn build(self) -> Result<ProductUpdateAction, ProductUpdateBuildError> {
+        let product_type = self.product_type.ok_or_else(|| {
+            ProductUpdateBuildError::MissingField("'product_type field is required".to_string())
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: missing second ' after product_type

Comment (agunde406) [reply sdk/src/protocol/product/payload.rs]: same in identifier

File: sdk/src/protocol/product/state.rs [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+            ProductBuildError::MissingField(_) => None,
+            ProductBuildError::EmptyVec(_) => None,
+        }
+    }
+}
+
+impl std::fmt::Display for ProductBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ProductBuildError::MissingField(ref s) => write!(f, "MissingField: {}", s),
```
Comment (peterschwarz) [inline sdk/src/protocol/product/state.rs]: Display message should be Human-readable, this is no better than just `Debug`.  Something like `"missing field \"{}\""` would be appropriate here.  Or with the EmptyVec error: `"\"{}\" must not be empty"`.

File: sdk/src/protocol/product/payload.rs [hunk 18]
```diff
@@ -0,0 +1,711 @@
...
+    }
+
+    pub fn with_identifier(mut self, identifier: String) -> ProductDeleteBuilder {
+        self.identifier = Some(identifier);
+        self
+    }
+
+    pub fn build(self) -> Result<ProductDeleteAction, ProductDeleteBuildError> {
+        let product_type = self.product_type.ok_or_else(|| {
+            ProductDeleteBuildError::MissingField("'product_type field is required".to_string())
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Missing second ' after product type

Comment (agunde406) [reply sdk/src/protocol/product/payload.rs]: same for identifier

File: sdk/protos/product_state.proto [hunk 19]
```diff
@@ -26,7 +26,7 @@ message Product {
   string identifier = 1;
 
   // What type of product is this (GS1)
-  ProductType type = 2;
+  ProductType product_type = 2;
```
Comment (agunde406) [inline sdk/protos/product_state.proto]: Can you leave a comment on the associated RFC mentioning this change?

File: contracts/product/src/handler.rs [hunk 20]
```diff
@@ -28,11 +28,15 @@ cfg_if! {
 }
 
 use grid_sdk::permissions::PermissionChecker;
-use grid_sdk::protos::product_payload::*;
-use grid_sdk::protos::product_state::Product;
+use grid_sdk::protocol::product::payload::{
```
Comment (agunde406) [inline contracts/product/src/handler.rs]: The commit message is a little long please take a look at https://chris.beams.io/posts/git-commit/

File: contracts/product/src/handler.rs [hunk 21]
```diff
@@ -100,47 +112,37 @@ impl ProductTransactionHandler {
             )));
         }
 
-        // Check that the agent has the pike permission "can_create_product" for the organization
-        check_permission(perm_checker, signer, "can_create_product")?;
+        // Check if product exists in state
```
Comment (agunde406) [inline contracts/product/src/handler.rs]: Please add comments into the commit message for any changes that are not directly replacing protos with native implementation. I may also make sense to break them up into their commits.

File: sdk/src/protocol/product/state.rs (hunk 14 above)
Comment (pa3ng) [reply sdk/src/protocol/product/state.rs]: Done ✅

File: sdk/src/protocol/product/state.rs (hunk 17 above)
Comment (pa3ng) [reply sdk/src/protocol/product/state.rs]: Done ✅

File: sdk/protos/product_payload.proto (hunk 11 above)
Comment (pa3ng) [reply sdk/protos/product_payload.proto]: Done ✅

Comment (pa3ng) [reply sdk/protos/product_payload.proto]: Done ✅

File: contracts/product/src/handler.rs (hunk 21 above)
Comment (pa3ng) [reply contracts/product/src/handler.rs]: Done ✅

File: contracts/product/src/handler.rs (hunk 20 above)
Comment (pa3ng) [reply contracts/product/src/handler.rs]: Done ✅

File: sdk/protos/product_state.proto (hunk 19 above)
Comment (pa3ng) [reply sdk/protos/product_state.proto]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 18 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 16 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 15 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 13 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 12 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: contracts/product/src/handler.rs (hunk 9 above)
Comment (peterschwarz) [reply contracts/product/src/handler.rs]: Looks like it's moved earlier.

File: sdk/src/protocol/product/payload.rs (hunk 7 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: contracts/product/src/handler.rs (hunk 9 above)
Comment (adeebahmed) [reply contracts/product/src/handler.rs]: Makes sense to move earlier, if the agent does not have permission it's a waste to run those subsequent checks.

File: sdk/src/protocol/product/payload.rs [hunk 22]
```diff
@@ -0,0 +1,742 @@
...
+            )
+        })?;
+        Ok(bytes)
+    }
+}
+
+impl IntoProto<protos::product_payload::ProductUpdateAction> for ProductUpdateAction {}
+impl IntoNative<ProductUpdateAction> for protos::product_payload::ProductUpdateAction {}
+
+/// Builder used to create a ProductUpdateActionBuilder
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Change to "Builder used to create a ProductUpdateAction"

File: sdk/src/protocol/product/payload.rs [hunk 23]
```diff
@@ -0,0 +1,742 @@
...
+            )
+        })?;
+        Ok(bytes)
+    }
+}
+
+impl IntoProto<protos::product_payload::ProductDeleteAction> for ProductDeleteAction {}
+impl IntoNative<ProductDeleteAction> for protos::product_payload::ProductDeleteAction {}
+
+/// Builder used to create a ProductDeleteActionBuilder
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Change to "Builder used to create a ProductDeleteAction"

File: sdk/src/protocol/product/payload.rs [hunk 24]
```diff
@@ -0,0 +1,742 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::protocol::schema::state::{DataType, PropertyValueBuilder};
+    use std::fmt::Debug;
+
+    #[test]
+    fn test_product_create_builder() {
```
Comment (agunde406) [inline sdk/src/protocol/product/payload.rs]: Please add a comment above each test describing the test (same for the tests below)

File: sdk/src/protocol/product/state.rs [hunk 25]
```diff
@@ -0,0 +1,570 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::protocol::schema::state::{DataType, PropertyValueBuilder};
+    use std::fmt::Debug;
+
+    #[test]
+    fn test_product_builder() {
```
Comment (agunde406) [inline sdk/src/protocol/product/state.rs]: Please add a comments describing the test here and the ones below

File: sdk/src/protocol/product/payload.rs (hunk 22 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 23 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/payload.rs (hunk 24 above)
Comment (pa3ng) [reply sdk/src/protocol/product/payload.rs]: Done ✅

File: sdk/src/protocol/product/state.rs (hunk 25 above)
Comment (pa3ng) [reply sdk/src/protocol/product/state.rs]: Done ✅

File: sdk/protos/product_state.proto [hunk 26]
```diff
@@ -26,13 +26,13 @@ message Product {
   string identifier = 1;
```
[the remark this answers is not in the document]
Comment (pa3ng) [reply sdk/protos/product_state.proto]: This change has been made (see `sdk/protos/product_state.proto`)
